anagram.1
./" Manpage for your_function
.TH ANAGRAM 1 "03 March 2024" "0.0"
.
.
.
.SH NAME
anagram.py \(en generates a full list of anagram sets of single words
in a chosen (combination of) language(s).
.
.
.
.SH SYNOPSIS
anagram.py  [\fIoptions\fP]  [\fIword\v'+0.2v'\s-31\s+3\v'-0.2v'\fP [ ... \fIword\v'+0.2v'\s-1n\s+1\v'-0.2v'\fP]]
.
.
.
.SH DESCRIPTION
anagram.py is a program that generates all sets of at least two (2) words matching
as mutual anagrams in the chosen language(s) present on the system.
All words are ordered such that every such common anagram set appears on its own line.
.LP
N.B.: only \fIsingle-word anagrams\fP are generated by this program,
no \fImulti-word anagrams\fP.
(For latter functionality, refer to the more advanced program `anagram\fBs\fP.py'.)
.LP
The \fIword\v'+0.2v'\s-31\s+3\v'-0.2v'\fP [ ... \fIword\v'+0.2v'\s-1n\s+1\v'-0.2v'\fP]
arguments are optional, and may be existing as well as non-existing words.
The program will in that case filter the results to only show the anagram words(s)
that match the given (combination of) word(s).
.LP
The program acts case-insensitive.
Accented characters are treated as equivalent to non-accented characters.
Interpunction symbols such as dots, apostrophs and hyphens do not need to match
in order to be part of an anagram.
.LP
Additionally, filters can be set for:
.IP \(en 3
word length, being the number of alphanumerical characters per word;
.IP \(en 3
minimal and/or maximal number of anagrams per set;
.IP \(en 3
characters to be all included and/or excluded.
.LP
Besides any single language, any \fIcombination\fP of languages can be set as well.
If no language option is given, Dutch is the default language.
.LP
The results are sent to standard output and can be piped to e.g. `less'
or other utilities and applications.
.LP
Perequisite is presence on the system of a word list in flat text format
of at least one language.
.LP
In its present form,
the program code references following language word lists: 
.IP \(bu 3
\fI/usr/share/dict/dutch\fP
.IP \(bu 3
\fI/usr/share/dict/american-english\fP
.IP \(bu 3
\fI/usr/share/dict/british-english\fP
.IP \(bu 3
\fI/usr/share/hunspell/de_DE_frami.dic\fP
.IP \(bu 3
\fI/usr/share/dict/french\fP
.IP \(bu 3
\fI/usr/share/dict/spanish\fP
.IP \(bu 3
\fI/usr/share/dict/italian\fP
.LP
.
.
.
.SH OPTIONS
anagram.py takes the following options:
.IP "\fB-a\fP"
Use American-English language dictionary.
.IP "\fB-b\fP"
Use British-English language dictionary.
.IP "\fB-d\fP"
Use Dutch language dictionary.
.IP "\fB-f\fP"
Use French language dictionary.
.IP "\fB-g\fP"
Use German language dictionary.
.IP "\fB-h\fP"
Help (this output).
.IP "\fB-i\fP"
Use Italian language dictionary.
.IP "\fB-s\fP"
Use Spanish language dictionary.
.IP "\fB-c\fP"
Use all language dictionaries combined.
.IP "\fB-l \fILENGTH\fP"
Print anagram sets with words of at least \fILENGTH\fP only,
being any positive integer representing the number of alphanumeric characters.
.IP "\fB-m \fIQTY\fP"
Print anagram sets with at least \fIQTY\fP results only,
being any positive integer.
If this option is not used, the default number of results per set is two (2).
.IP "\fB-M \fIQTY\fP"
Print anagram sets with at most \fIQTY\fP results only, being a positive integer.
.IP "\fB-I \fICHARS\fP"
Include words with all of these \fICHARS\fP only.
This option does take character case, accent marks and interpunction literally.
.IP "\fB-x \fICHARS\fP"
Exclude words with any of these \fICHARS\fP.
This option does take character case, accent marks and interpunction literally.
.LP
All options can be combined.
If no language option is given, Dutch is the default language.
.
.
.
.SH COMPATIBILITY
Python3 should be installed.
In order to be able to use (a) certain language option(s),
the associated language file(s) as specified above should be present on the system as well.
.
.
.SH BUGS
No known bugs.
.
.
.SH AUTHOR
Written by Rob Toscani (rob_toscani@yahoo.com)
.
.
.
.SH REPORTING BUGS
Please report any bugs to the author by e-mail or via https://github.com/jazzfan2/anagram/issues


anagram.sh
#!/bin/bash
# Name       : anagram.sh
# Author     : Rob Toscani
# Date       : 22-02-2024
# Description: Shell script based on awk, generating a full list of anagrams in the 
# chosen languages present on the system.
# In case a (combination of) word(s) - whether or not existing(!) - is given as 
# argument(s), it only generates the anagrams that fit the (combination of) word(s). 
# Additionally, filters can be set for:
# - anagram-length
# - minimal and/or maximal number of anagrams per solution
# - characters to be all included and/or excluded
# Besides any single language, any *combination* of languages can be set as well.
# Output can be piped to e.g. 'less' or other utilities and applications.
#
# Examples:
#
#   anagram.sh -abdgsi | sed -r 's/ {2,}/\t/g' | grep -E "[^\t]\t+[^\t]" | allign_tsv.sh -
#   anagram.sh -abdgsi | sed -r 's/ {2,}/\t/g' | grep -E "[^\t]\t+[^\t]" | text2troff - | trofms
#   anagram.sh -abdgsi | sed 's/ *$//g' | sed -r 's/ {2,}/\t/g' | grep -E "[\t].+[\t]" | text2troff -s - | trofms
#
# Perequisite is presence on the system of a word list in flat text format of at least
# one language.
#
# Alternative version of anagram, using a real 2D array instead of a 'quasi-2D'-array.
# Same speed.
#
######################################################################################
#
# Copyright (C) 2024 Rob Toscani <rob_toscani@yahoo.com>
#
# anagram.sh is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# anagram.sh is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
######################################################################################
#
set -o noclobber    # Prevent files to be accidentally overwritten by standard output redirection


if [[ -d /tmp/ramdisk/ ]]; then                          # tmpfs
    tmpfiledir="/tmp/ramdisk"
elif [[ -d /dev/shm/ ]]; then
    tmpfiledir="/dev/shm"
else
    tmpfiledir="."
fi

dictionary_nl="/usr/share/dict/dutch"
dictionary_am="/usr/share/dict/american-english"
dictionary_br="/usr/share/dict/british-english"
dictionary_de="/usr/share/hunspell/de_DE_frami.dic"
dictionary_fr="/usr/share/dict/french"
dictionary_sp="/usr/share/dict/spanish"
dictionary_it="/usr/share/dict/italian"
dict="$tmpfiledir/dict$RANDOM"


process_dict_de()
{
    grep -v ^# $dictionary_de 2>/dev/null | sed 's/\/[^/]*//g' | sort | uniq
}


helptext()
# Text printed if -h option (help) or a non-existing option has been given:
{
    while read "line"; do
        echo "$line" >&2         # print to standard error (stderr)
    done << EOF

Usage:
anagrams.sh [-abcdfghislmMIx] [WORD(1) [ ... WORD(n)]]

-a	American-English
-b	British-English
-d	Dutch
-f	French
-g	German
-h	Help (this output)
-i	Italian
-s	Spanish
-c	All languages combined
-l LENGTH
|	Print solutions with word LENGTH only
-m QTY
|	Print solutions with at least QTY anagrams
-M QTY
|	Print solutions with at most QTY anagrams
-I CHARS
|	Include words with all of these CHARS
-x CHARS
|	Exclude words with any of these CHARS
EOF
}


default=$dictionary_nl
word_args=""
qty_min=2         # Anagram means at least two matching words, so this is the default
qty_max=100
filterlength=0
incl_chars="."    # Default: dot means include all characters
excl_chars="_"    # Default: underscore does not appear so can always be excluded
count=0
touch $dict

while getopts "abcdfghisl:m:M:I:x:" OPTION; do
    case $OPTION in
        a) cat $dictionary_am >> $dict; (( count += 1 )) ;;
        b) cat $dictionary_br >> $dict; (( count += 1 )) ;;
        d) cat $dictionary_nl >> $dict; (( count += 1 )) ;;
        f) cat $dictionary_fr >> $dict; (( count += 1 )) ;;
        g) process_dict_de    >> $dict; (( count += 1 )) ;;
        h) helptext; exit 0 ;;
        i) cat $dictionary_it >> $dict; (( count += 1 )) ;;
        s) cat $dictionary_sp >> $dict; (( count += 1 )) ;;
        c) process_dict_de    >| $dict
           cat $dictionary_nl $dictionary_am \
               $dictionary_br $dictionary_fr \
               $dictionary_sp $dictionary_it >> $dict; count=7 ;;
        l) filterlength=$OPTARG ;;
        m) qty_min=$OPTARG ;;
        M) qty_max=$OPTARG ;;
        I) incl_chars=$OPTARG ;;
        x) excl_chars=$OPTARG ;;
        *) helptext; exit 1 ;;
    esac
done
shift $((OPTIND-1))

[[ $# != 0 ]] && word_args="$@"                     # Word(s) argument(s)

([[ $# != 0 ]] && (( qty_min == 2 ))) && qty_min=1  # In case just 1 anagram matches *non-existent* word(s) 

chargrep=""
for (( i = 0; i < ${#incl_chars}; i += 1 )); do
    chargrep=$chargrep" | grep -i ${incl_chars:$i:1}"
done
chargrep=${chargrep:2}


echo -e "One moment, the output is being prepared ...\n" >&2

[[ -s $dict ]] && default=$dict

if (( count > 1 )); then      # If minimally 2 dictionaries are combined: apply 'sort' en 'uniq'
    sort $default | uniq
else
    cat $default
fi |

awk -v qty_min=$qty_min -v qty_max=$qty_max -v filterlength=$filterlength \
    -v word_args="$word_args" '\
    function normalize(string)
    {
         string = tolower(string)
         gsub(/[áàäâå]/, "a", string)        # Normalize all characters to lower case and w/o accent marks
         gsub(/[éèëê]/,  "e", string)
         gsub(/[ïíì]/,   "i", string)
         gsub(/[óòöôø]/, "o", string)
         gsub(/[úùü]/,   "u", string)
         gsub(/[ñ]/,     "n", string)
         gsub(/[ç]/,     "c", string)

         gsub(/['\"''\'' .&-]/, "", string)  # Remove other non-alphanumeric characters

         split(string,chars,"")              # Array of all the string`s normalized characters
         signature = ""
         for (i in chars){
             signature = signature chars[i]
         }
         return signature                    # Unique sorted character combination
    }

    BEGIN {
         # Setting for looping through an array in ascending index order, see:
         # https://www.gnu.org/software/gawk/manual/gawk.html#Controlling-Scanning

         PROCINFO["sorted_in"]="@val_str_asc"
         split("",anagrams)

         word_args = normalize(word_args)    # Get signature of (optional) word(s) argument(s)
     }

     {
         signature = normalize($0)                             # Get signature of present dictionary word
         anagrams[signature][length(anagrams[signature])] = $0 # 2D-array of all dict. words per signature 
     }

     END {
         for (signature in anagrams){
             if ((length(word_args) == 0 || signature == word_args) &&
                length(anagrams[signature]) >= qty_min && length(anagrams[signature]) <= qty_max &&
                (filterlength == 0 || filterlength == length(signature))){
                 for (i = 0; i < length(anagrams[signature]); i++){
                     if (length(signature) < 18)
                         printf("%-20s", anagrams[signature][i])
                     else
                         printf("%-40s", anagrams[signature][i])
                 }
                 printf("\n")
             }
         }
     }' | sort | grep -iv [$excl_chars] | eval "$chargrep"

